autocert/examples/hello-mtls/go-grpc/server/server.go

```diff
@@ -0,0 +1,139 @@
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/smallstep/certificates/autocert/examples/hello-mtls/go-grpc/hello"
+)
+
+const (
+	autocertFile  = "/var/run/autocert.step.sm/site.crt"
+	autocertKey   = "/var/run/autocert.step.sm/site.key"
+	autocertRoot  = "/var/run/autocert.step.sm/root.crt"
+	tickFrequency = 15 * time.Second
+)
+
+// Uses techniques from https://diogomonica.com/2017/01/11/hitless-tls-certificate-rotation-in-go/
+// to automatically rotate certificates when they're renewed.
+
+type rotator struct {
+	sync.RWMutex
+	certificate *tls.Certificate
+}
+
+func (r *rotator) getCertificate(hello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	r.RLock()
+	defer r.RUnlock()
+	return r.certificate, nil
+}
+
+func (r *rotator) loadCertificate(certFile, keyFile string) error {
+	r.Lock()
+	defer r.Unlock()
+
+	c, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		return err
+	}
+
+	r.certificate = &c
+
+	return nil
+}
+
+func loadRootCertPool() (*x509.CertPool, error) {
+	root, err := ioutil.ReadFile(autocertRoot)
+	if err != nil {
+		return nil, err
+	}
+
+	pool := x509.NewCertPool()
+	if ok := pool.AppendCertsFromPEM(root); !ok {
+		return nil, errors.New("Missing or invalid root certificate")
+	}
+
+	return pool, nil
+}
+
+// Greeter is a service that sends greetings.
+type Greeter struct{}
+
+// SayHello sends a greeting
+func (g *Greeter) SayHello(ctx context.Context, in *hello.HelloRequest) (*hello.HelloReply, error) {
+	return &hello.HelloReply{Message: "Hello " + in.Name}, nil
```

Comment:
Looks like `in.Name` is coming from the protobuf request? In the HTTPS example I was intentionally trying to demonstrate pulling the name from a validated certificate because it'd be the next step if you wanted to do authorization... so basically that example gets you to the point where you have authenticated identity and are ready to implement your own access control. Might be worthwhile to do the same thing here. If I understand correctly, gRPC has some abstraction around this...

Reply:
Fixed with [64cbac4](/smallstep/certificates/pull/33/commits/64cbac4e8173f556c7e66ce1d40bdbdbc8221c5d)